drivers: intc: plic: add shell cmd to get irq stats for debugging

ycsin opened the pull request:
Introduced `CONFIG_PLIC_SHELL` to enable the build of shell debugging command to get the hit count of each interrupt controller's IRQ line. This is especially useful when working with dynamically installed ISRs, which will be the case for `plic_sw`.

Example usage:

```
uart:~$ plic stats get interrupt-controller@c000000
   IRQ        Hits
==================
    10         177

uart:~$ plic stats get interrupt-controller@c000000
   IRQ        Hits
==================
    10         236

uart:~$ plic stats clear interrupt-controller@c000000
Cleared stats of interrupt-controller@c000000.

uart:~$ plic stats get interrupt-controller@c000000
   IRQ        Hits
==================
    10          90

```

ycsin committed:
drivers: intc: plic: add shell cmd to get irq stats for debugging

Introduced `CONFIG_PLIC_SHELL` to enable the build of shell
debugging command to get the hit count of each interrupt
controller's IRQ line. This is especially useful when working
with dynamically installed ISRs, which will be the case for
`plic_sw`.

Example usage:

```
uart:~$ plic stats get interrupt-controller@c000000
   IRQ        Hits
==================
    10         177

uart:~$ plic stats get interrupt-controller@c000000
   IRQ        Hits
==================
    10         236

uart:~$ plic stats clear interrupt-controller@c000000
Cleared stats of interrupt-controller@c000000.

uart:~$ plic stats get interrupt-controller@c000000
   IRQ        Hits
==================
    10          90

```

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Maxim Adelman <imax@meta.com>

pdgendt commented on drivers/interrupt_controller/Kconfig.plic:
```diff
@@ -10,3 +10,14 @@ config PLIC
+
+if PLIC
+
+config PLIC_DEBUG
```
IMO we should keep consistent with `Kconfig` naming and user `_SHELL` for symbols that enable shell commands.

```suggestion
config PLIC_SHELL
```

fkokosinski commented on drivers/interrupt_controller/Kconfig.plic:
```diff
@@ -10,3 +10,14 @@ config PLIC
 	  Platform Level Interrupt Controller provides support
 	  for external interrupt lines defined by the RISC-V SoC.
+
+if PLIC
```
Hi! Thanks for this PR.

I'm wondering - could we generalize the PLIC shell to an Interrupt Controller shell, similar to what other drivers do? This would essentially mean that two new symbols would have to be implemented for each arch:
* `arch_irq_stat_get`
* `arch_irq_stat_reset`

And the generic driver in `drivers/interrupt_controller/intc_shell.c` would use them. Since this would be an opt-in feature, we wouldn't have to implement the support for it in every interrupt controller that is present in Zephyr right away.

ycsin replied:
Each of the interrupt controllers is quite unique and doesn't have a unified interface like `i2c`, `flash`... I'd rather keep this within PLIC for now

fkokosinski replied:
Fair enough :)

fkokosinski commented on drivers/interrupt_controller/Kconfig.plic:
```diff
@@ -10,3 +10,14 @@ config PLIC
+	depends on SHELL
+	help
+	  Enable additional shell commands useful for debugging.
+	  Caution: This will use a lot of RAM (IRQs * sizeof(size_t)).
```
If that's the case, do we want to print a warning from CMake if this symbol is enabled, so that we're sure the user is informed about this?

cfriedt commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -337,6 +356,117 @@ static int plic_init(const struct device *dev)
+static int cmd_stats(const struct shell *sh, size_t argc, char *argv[])
+{
+	const struct device *dev;
+	int ret = parse_device(sh, argc, argv, &dev);
```
Maybe as a follow-up PR (or can be done here too), would be nice to be able to specify some additional arguments.

Positional arguments could just be irqs (so print only the named irqs, otherwise all).

`plic stats get <device> [irq..]`

Additional non-positional arguments could be e.g `--min count` to only print irqs greater than a `count`.

A bit of a pipe dream, but it would be nice to print the human readable device ids associated with each irq (e.g. `--name` or similar option). I know labels went the way of the dinosaur, but maybe there is some other way.

`plic stats get <device> --name`

Irq, name, count

ycsin replied:
I've added an optional argument to only print irqs greater than a `count`
```
plic stats get <device> [minimum hits]
```

cfriedt reviewed (commented):
If there is another respin, we should fix the implicit cast (violates Misra), and also some suggestions for features

cfriedt commented:
Maybe an improvement (would allow for tab completion)

`plic stats get <device>`
`plic stats clear <device>`

Rather than
`plic stats <device>`
`plic clear_stats <device>`

ycsin committed:
drivers: intc: plic: brackets for if-conds & use explicit comparison

if-conditionals should have brackets according to Zephyr's
coding standard, and explicitly compares `edge_irq` against 0.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -337,6 +356,117 @@ static int plic_init(const struct device *dev)
+	const struct device *dev;
+	int ret = parse_device(sh, argc, argv, &dev);
+
+	if (ret) {
```
fixed

cfriedt reviewed (approved):
@fkokosinski - please give it a spin as well. I'm ok with it as-is. Eventually it would be nice to make this a part of some future generic interrupt controller device API.
